File: framework/languages/java/e2e/ChannelEgressRouting/Role/src/main/java/systems/zlink/e2e/channelegress/role/Config12Spot.java
package systems.zlink.e2e.channelegress.role;

import java.time.Duration;
import java.util.ArrayList;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.e2e.channelegress.shared.Contracts;
import systems.zlink.e2e.channelegress.shared.EvidenceState;
import systems.zlink.framework.actors.ZLinkActor;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkSpot;
import systems.zlink.framework.spots.ZLinkSpotContext;
import systems.zlink.framework.spots.ZLinkSpotCreateResponse;
import systems.zlink.framework.spots.ZLinkTimer;

public final class Config12Spot implements ZLinkSpot<ZLinkActor> {
    private final ZLinkSpotContext context;
    private final EvidenceState evidence;
    private ZLinkTimer timer;

    public Config12Spot(ZLinkSpotContext context, EvidenceState evidence) {
        this.context = context;
        this.evidence = evidence;
    }

    @Override
    public ZLinkSpotContext context() {
        return context;
    }

    @Override
    public void configure() {
        context.handlers().addHandler(SpotWorkflowHandler.class);
    }

    @Override
    public CompletionStage<ZLinkSpotCreateResponse> onCreate(ZLinkMessage ignored) {
        evidence.add("spot-initialize", "spot=" + context.spotId());
        return CompletableFuture.completedFuture(ZLinkSpotCreateResponse.accept());
    }

    @Override
    public CompletionStage<Void> onJoinedActor(ZLinkActor ignored) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onLeaveActor(ZLinkActor ignored) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onClosing() {
        evidence.add("spot-closing", "spot=" + context.spotId());
        return CompletableFuture.completedFuture(null);
    }

    CompletionStage<Void> startTimer(String name) {
        return context.addTimer(
                name,
                Duration.ofMillis(1),
                SpotWorkflowTimerHandler.class,
                null)
            .thenAccept(value -> timer = value);
    }

    void closeTimer() {
        ZLinkTimer current = timer;
        if (current != null) {
            current.close();
            timer = null;
        }
    }
}

File: framework/languages/node/e2e/ChannelEgressRouting/Role/Spots/config12-spot.ts
import { Injectable, Scope } from '@nestjs/common';
import type { ZLinkInstanceSpot, ZLinkInstanceSpotContext } from '@zlink-systems/framework';
import { SpotWorkflowHandler } from '../Handlers/spot-handlers';
import { EvidenceStore } from '../Support/evidence-store';

@Injectable({ scope: Scope.TRANSIENT })
export class Config12Spot implements ZLinkInstanceSpot {
  readonly context!: ZLinkInstanceSpotContext;

  constructor(private readonly evidence: EvidenceStore) {}

  configure(): void {
    this.context.handlers.addPacket(SpotWorkflowHandler);
  }

  async onInitialize(): Promise<void> {
    this.evidence.add(`spot-initialize|rid=${this.evidence.rid}|spot=${this.context.spotId}`);
  }

  async onClosing(): Promise<void> {
    this.evidence.add(`spot-closing|rid=${this.evidence.rid}|spot=${this.context.spotId}`);
  }
}
